packages/headless/doc-parser/src/entity-sorter.ts

```diff
@@ -0,0 +1,55 @@
+import {
+  AnyEntity,
+  Entity,
+  EntityWithTypeAlias,
+  FuncEntity,
+  ObjEntity,
+} from './entity';
+
+export function sortEntities(entities: AnyEntity[]) {
+  const [methods, attributes] = divideIntoGroups(entities, (entity) =>
+    entity.kind === 'function' ? SortGroup.First : SortGroup.Second
+  ) as [FuncEntity[], (Entity | EntityWithTypeAlias | ObjEntity)[]];
+
+  const [optionalAttributes, mandatoryAttributes] = divideIntoGroups(
+    attributes,
+    (entity) => (entity.isOptional ? SortGroup.First : SortGroup.Second)
+  );
+
+  const sortedMandatoryAttributes = alphabeticallySortEntities(
+    mandatoryAttributes
+  );
+  const sortedOptionalAttributes = alphabeticallySortEntities(
+    optionalAttributes
+  );
+  const sortedMethods = alphabeticallySortEntities(methods);
+
+  return [
+    ...sortedMandatoryAttributes,
+    ...sortedOptionalAttributes,
+    ...sortedMethods,
+  ];
+}
+
+enum SortGroup {
+  First,
+  Second,
+}
```

Comment:
Assuming we keep `divideIntoGroups` method, since there are only two groups, it would be possible to simplify to just a boolean.

Reply:
I'm also adding a comment to clarify what the boolean value means, then.